File: 2ndSemester/kursprogramowania/lista1/cpp/src/primes.cpp
#include <primes.hpp>

#include <string>
#include <cmath>
#include <vector>


std::vector<bool> Primes::sieve(const int n) noexcept(false)
{
    if (n <= 1)
    {
        throw std::invalid_argument(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "parameter n should be greater than 1, got " + std::to_string(n));
    }
    
    std::vector<bool> primes = {false, false};

    for (int i = 2; i <= n; i++)
    {
        primes.push_back(true);
    }

    for (int i = 2; i <= static_cast<int>(std::floor(std::sqrt(static_cast<double>(n)))); i++)
    {
        if(primes[static_cast<size_t>(i)])
        {
            for (int j = i * i; j <= n; j += i)
                primes[static_cast<size_t>(j)] = false;
        }
    }
    
    return primes;
}

std::vector<int> Primes::primesList(const int n) noexcept(false)
{
    if (n <= 1)
    {
        throw std::invalid_argument(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "parameter n should be greater than 1, got " + std::to_string(n));
    }

    std::vector<int> primesList;
    std::vector<bool> isPrime = Primes::sieve(n);

    for (int i = 2; i <= n; i++)
    {
        if (isPrime[static_cast<size_t>(i)])
        {
            primesList.push_back(i);
        }
    }
    return primesList;
    
}

File: 2ndSemester/kursprogramowania/lista1/java/Primes.java
import java.util.ArrayList;

/**
 * @class Primes
 * 
 * @brief Primes class is a statitc class containing only statis methods related to prime numbers
 */
public final class Primes 
{
    private Primes() throws InstantiationError
    {
        throw new InstantiationError("This is a static class");
    }

    /**
     * @brief creates a vector of booleans, for each number from 0 to n tells if the number is prime
     * 
     * @param n - any number, but for numbers <= 1 an exception will be raised
     * @return std::vector<bool> - true for a prime, false otherwise 
     */
    public static ArrayList<Boolean> sieve(final int n) throws IllegalArgumentException
    {
        if (n <= 1) 
        {
            throw new IllegalArgumentException("parameter n should be greater than 1, got " + n);
        }

        ArrayList<Boolean> primes = new ArrayList<Boolean>();
        primes.add(false);
        primes.add(false);

        for (int i = 2; i <= n; i++)
        {
            primes.add(true);
        }

        for (int i = 2; i <= Math.floor(Math.sqrt(n)); i++)
        {
            if (primes.get(i)) 
            {
                for (int j = i * i; j <= n; j += i)
                {
                    primes.set(j, false);
                }
            }
        }
        return primes;
    }

    /**
     * @brief creates a vector of primes between 2 and n
     * 
     * @param n - any number, but for numbers <= 1 an exception will be raise
     * @return std::vector<int> containing primes
     */
    public static ArrayList<Integer> primesList(final int n) throws IllegalArgumentException
    {
        if (n <= 1) 
        {
            throw new IllegalArgumentException("parameter n should be greater than 1, got " + n);
        }

        ArrayList<Integer> plist = new ArrayList<Integer>();
        ArrayList<Boolean> isPrime = sieve(n);

        for (int i = 2; i <= n; i++)
        {
            if (isPrime.get(i))
            {
                plist.add(i);
            }
        }
        
        return plist;
    }
}
